lib/unifex/code_generator/base_type.ex

```diff
@@ -99,69 +101,88 @@ defmodule Unifex.CodeGenerator.BaseType do
 
   Returns an empty string if the type does not provide destructor
   """
-  @spec generate_destruction(t, name :: atom, module) :: CodeGenerator.code_t()
-  def generate_destruction(type, name, code_generator) do
-    call(type, :generate_destruction, [name], code_generator)
+  @spec generate_destruction(t, name :: atom, module, ctx :: map) :: CodeGenerator.code_t()
+  def generate_destruction(type, name, code_generator, ctx) do
+    call(type, :generate_destruction, [name], code_generator, ctx)
   end
 
   @doc """
   Generates parsing of UNIFEX_TERM `argument` into the native variable
   """
-  def generate_arg_parse(type, name, argument, postproc_fun \\ & &1, code_generator) do
+  def generate_arg_parse(type, name, argument, postproc_fun \\ & &1, code_generator, ctx) do
+    ctx =
+      ctx
+      |> Map.put(:postproc_fun, postproc_fun)
+
     call(
       type,
       :generate_arg_parse,
       [argument, name],
       code_generator,
-      %{postproc_fun: postproc_fun}
+      ctx
     )
     |> postproc_fun.()
   end
 
-  def generate_arg_name(type, name, code_generator) do
-    generate_native_type(type, code_generator)
+  def generate_arg_name(type, name, code_generator, ctx) do
+    generate_native_type(type, code_generator, ctx)
     |> Bunch.listify()
     |> Enum.map(fn
       {_type, suffix} -> ~g<#{name}#{suffix}>
       _type -> ~g<#{name}>
     end)
   end
 
-  def generate_native_type(type, mode \\ :default, code_generator) do
-    call(type, :generate_native_type, [], code_generator, %{mode: mode})
+  def generate_native_type(type, mode \\ :default, code_generator, ctx) do
+    ctx =
+      ctx
+      |> Map.put(:mode, mode)
+
+    call(type, :generate_native_type, [], code_generator, %{ctx | mode: mode})
   end
 
-  defp call(full_type, callback, args, code_generator, ctx \\ %{}) do
+  defp call(full_type, callback, args, code_generator, ctx) do
     {type, subtype} =
       case full_type do
         {type, subtype} -> {type, subtype}
         type -> {type, nil}
       end
 
+    ctx =
+      with %{customized_types: %{^type => type_spec}} <- ctx do
+        ctx
+        |> Map.put(:type_spec, type_spec)
+      else
+        _ -> ctx
+      end
+
     module =
-      Module.concat(Unifex.CodeGenerator.BaseTypes, type |> to_string() |> String.capitalize())
+      with %{customized_types: %{^type => _type_spec}} <- ctx do
+        BaseTypes.Struct
+      else
+        _ -> Module.concat(BaseTypes, type |> to_string() |> String.capitalize())
+      end
 
     gen_aware_module = Module.concat(module, code_generator)
 
-    default_gen_aware_module =
-      Module.concat(Unifex.CodeGenerator.BaseTypes.Default, code_generator)
+    default_gen_aware_module = Module.concat(BaseTypes.Default, code_generator)
 
     args =
-      args ++ [Map.merge(%{generator: code_generator, type: full_type, subtype: subtype}, ctx)]
+      args ++ [Map.merge(ctx, %{generator: code_generator, type: full_type, subtype: subtype})]
```

Comment:
Old version of code was working properly, because contexts hadn't been passed in every call of public functions from `BaseType` module. Since I have changed it, it's necessary to change this order, in due to override existing values in `ctx` - before my changes `ctx` hadn't contain these keys